contribute/design/3.x/cstor_replica_placement/placement.md

```diff
@@ -0,0 +1,111 @@
+    - [User Stories](#user-stories)
+      - [Support to disable overprovisioning on cstor pool](#support-to-disable-overprovisioning-on-cstor-pool)
+    - [Proposed Implementation](#proposed-implementation)
+      - [cvr-scheduler](#cvr-scheduler)
+    - [Steps to perform user stories](#steps-to-perform-user-stories)
+    - [Schema changes](#schema-changes)
+
+## Summary
+
+This proposal brings out the design details to implement volume placement on
+CStor pool.
+
+## Motivation
+
+In current impementation, the CVR (CStor Volume Replica) gets bound to a CSPI
+(CStor Pool Instance) by the cvr controller in a random fashion. As a result of
+this, users have no way to specify to bind a CVR to a preferred CSPI. 
+
+- CStor volume placement should be supported for the following cases
+(not exhaustive):
+  1. Disable overprovisoning on CSPI
+  2. Migration of a CVR from one CSPI to another
+  3. Selecting the best CSPI(s) for the CVR(s) based on node and zone locality
+
+
+### Goals
+
+- The bind of a CVR to a CSPI should be deterministic and users should be able to
+specify policies for the binding. As a first cut implementation, the policy of 
+interest here is the overprovisioning restriction. 
+
+### Non-Goals
+
+- Migrating a CVR from one CSPI to another
+- Policies around node and zone locality
+
+## Proposal
+
+### User Stories
+
+#### Disable overprovisioning on CStor pools
+As an OpenEBS user, I should be able to disable overprovisioning on CStor pools.
+
+#### Migrating a CVR from one CSPI to another
+As an OpenEBS user, I should be able to migrate volume from one CStor pool to
+another CStor pool.
+
+#### Enforce policies on CVR binding
+As an OpenEBS user, I should be able to enforce policies e.g. node,zone locality
+to bind to a most suitable(best effort) CSPI.
+
+### Proposed Implementation
+
+#### CVR-Scheduler
+
+In current impementation, the CVR (CStor Volume Replica) gets bound to a CSPI
+(CStor Pool Instance) by the cvr controller in a random fashion. See this [link]
+(https://github.com/openebs/cstor-operators/blob/master/pkg/controllers/cstorvolumeconfig/volume_operations.go#L396)
+
+A separate controller i.e cvr-scheduler watches for CVR(s). Once the CVR is 
+created, it is unassigned, meaning it does not belong to any CSPI. Now, this 
+cvr-scheduler enforces policies and requirements e.g. overprovisioning etc and 
+picks up appropriate or most suitable CSPI for the particular CVR and assigns 
+this CVR to a CSPI by patching the CVR. This patching will basically put the 
+selected CSPI name and UUID on the CVR.
+
+The replica controller that also watches for the CVRs. It ignores the CVR that 
+has not been assigned any CSPI. Once the CVR is assigned to a CSPI, the replica 
+controller processes the CVR to carry out the volume provisioning.
+
+In case a quick binding is required, there will be an option to by pass the cvr
+scheduler and the usual way of randomised binding will prevail via the replica
+controller.
```

Comment:
How users can pass the policy options? Is it via StorageClass parameters?

Reply:
Ok, Seen in the below lines.
> The policies for the cvr-scheduler (e.g. overprovisioning, binding) can be passedto the CVR(s) via the CSPC.

[Story-1:]
Are all the policies are specified via CSPC? If yes, then I have one followup question, I am a user of OpenEBS cStor and I have >=6 cStor pools(On different nodes) and I want to place my important workload in pools where volume replica count is less (Or) placing replicas in pool based on available capacity. IMO this kind of policies should be passed via storageclass since it is related to volume replica placement.


[Story-2]:
If I am not wrong as of now we are interested only in [ThickProvisioning](https://github.com/openebs/api/blob/master/pkg/apis/cstor/v1/cstorpoolcluster.go#L111-L113)(OverProvisioing). IMO that it is not a policy but it is a kind of parameter that specifies the property/nature of the pool. If it is specified then we will not allow placement of replicas in the pool if available capacity is less than the current volume replica capacity, right?

Reply:
With Re to [Story-1]: 
Passing policies via storage class makes sense to me. I think we should have two ways to specify policies : 
1. via CSPC ( To enforce pool level constraint and policies e.g. `overprovisioning`
2. via StorageClass ( To select most unused CStorPool )
Will make the changes and commit.

Reply:
With Re to [Store-2] : 
My thoughts are the following: 
We should not pull in the data layer concepts to the control plane. The Control plane just knows how to orchestrate, that's all. 
In that sense, `overprovisioning` is a mere policy, although it is a low-level property at CStor engine too. 
Plus, there is no need to get/fetch the pool's available capacity. We just need to know the total usable capacity of the pool and how many total replicas are present on the pool that request some amount of capacity. Then we calculate the remaining capacity that is guaranteed to be available for further placement. 
The catch is whenever a volume replica is placed on a pool with overprovisioning disabled, the data layer property is also set so that the replicas do not use more than requested. 

Note: Fecthcing available capacity for this kind of requirements is erroneous in my opinion due to the fact the I/O can happen in parallel. 
Correct me if I am missing something. Open to discussion.

Reply:
> there is no need to get/fetch the pool's available capacity. We just need to know the total usable capacity of the pool and how many total replicas are present on the pool that request some amount of capacity.

Yeah, make sense. I was under a different mindset thinking that we are populating correct usable capacity on CSPI status but it is not.

> Note: Fecthcing available capacity for this kind of requirements is erroneous in my opinion due to the fact the I/O can happen in parallel.

It is not recommended to query very frequently about zfs properties since it will degrade the performance to some more extent.